core/src/main/java/org/infinispan/container/offheap/BoundedOffHeapDataContainer.java

```diff
@@ -73,103 +73,106 @@ protected void entryReplaced(long newAddress, long oldAddress) {
       try {
-         long lruNode = offHeapEntryFactory.getLruNode(oldAddress);
          if (trace) {
-            log.tracef("Replacing LRU node: 0x%016x. OldValue: 0x%016x NewValue: 0x%016x", lruNode, oldAddress,
+            log.tracef("Replacing LRU node: OldValue: 0x%016x NewValue: 0x%016x", oldAddress,
                        newAddress);
          }
-         // We have to update the lru node to point to the new address and vice versa
-         offHeapEntryFactory.setLruNode(newAddress, lruNode);
-         OffHeapLruNode.setEntry(lruNode, newAddress);
 
-         moveToEnd(lruNode);
+         removeNode(oldAddress);
+         addEntryAddressToEnd(newAddress);
 
          currentSize += newSize;
          currentSize -= oldSize;
+         // Must only remove entry while holding lru lock now
+         super.entryReplaced(newAddress, oldAddress);
       } finally {
          lruLock.unlock();
       }
-      super.entryReplaced(newAddress, oldAddress);
    }
 
    @Override
    protected void entryCreated(long newAddress) {
-      int hashCode = offHeapEntryFactory.getHashCode(newAddress);
       long newSize = sizeCalculator.applyAsLong(newAddress);
       lruLock.lock();
       try {
          currentSize += newSize;
-         addEntryAddressToEnd(newAddress, hashCode);
+         addEntryAddressToEnd(newAddress);
+         super.entryCreated(newAddress);
       } finally {
          lruLock.unlock();
       }
-      super.entryCreated(newAddress);
    }
 
    @Override
    protected void entryRemoved(long removedAddress) {
       long removedSize = sizeCalculator.applyAsLong(removedAddress);
-      long lruNode = offHeapEntryFactory.getLruNode(removedAddress);
-      assert lruNode != 0;
       lruLock.lock();
       try {
          // Current size has to be updated in the lock
-         currentSize -= removedSize;
-         boolean middleNode = true;
-         if (lruNode == lastAddress) {
-            if (trace) {
-               log.tracef("Removing last LRU node at 0x%016x", lruNode);
-            }
-            long previousLRUNode = OffHeapLruNode.getPrevious(lruNode);
-            if (previousLRUNode != 0) {
-               OffHeapLruNode.setNext(previousLRUNode, 0);
-            }
-            lastAddress = previousLRUNode;
-            middleNode = false;
+         currentSize -=  removedSize;
+         removeNode(removedAddress);
+         // Removals are only done while holding lru lock now
+         super.entryRemoved(removedAddress);
+      } finally {
+         lruLock.unlock();
+      }
+   }
+
+   /**
+    * Removes the address node and updates previous and next lru node pointers properly
+    * The {@link BoundedOffHeapDataContainer#lruLock} <b>must</b> be held when invoking this
+    * @param address
+    */
+   private void removeNode(long address) {
+      boolean middleNode = true;
+      if (address == lastAddress) {
+         if (trace) {
+            log.tracef("Removed LRU node at end 0x%016x", address);
```

Comment:
Since we don't have a LRU node "instance" any more, I prefer talking about the LRU list and the entry, e.g. "Removed entry 0x%016x from the end of the LRU list".

Reply:
Sure